temporalio/worker/_nexus.py

```diff
@@ -0,0 +1,467 @@
+            return temporalio.api.nexus.v1.Failure(
+                message=_api_failure.pop("message", ""),
+                metadata={"type": "temporal.api.failure.v1.Failure"},
+                details=json.dumps(_api_failure).encode("utf-8"),
+            )
+        except BaseException as err:
+            return temporalio.api.nexus.v1.Failure(
+                message=f"{err.__class__.__name__}: {err}",
+                metadata={"type": "temporal.api.failure.v1.Failure"},
+            )
+
+    async def _operation_error_to_proto(
+        self,
+        err: nexusrpc.OperationError,
+    ) -> temporalio.api.nexus.v1.UnsuccessfulOperationError:
+        return temporalio.api.nexus.v1.UnsuccessfulOperationError(
+            operation_state=err.state.value,
+            failure=await self._exception_to_failure_proto(err),
+        )
+
+    async def _handler_error_to_proto(
+        self, err: nexusrpc.HandlerError
+    ) -> temporalio.api.nexus.v1.HandlerError:
+        return temporalio.api.nexus.v1.HandlerError(
+            error_type=err.type.value,
+            failure=await self._exception_to_failure_proto(err),
+            retry_behavior=(
+                temporalio.api.enums.v1.NexusHandlerErrorRetryBehavior.NEXUS_HANDLER_ERROR_RETRY_BEHAVIOR_RETRYABLE
+                if err.retryable
+                else temporalio.api.enums.v1.NexusHandlerErrorRetryBehavior.NEXUS_HANDLER_ERROR_RETRY_BEHAVIOR_NON_RETRYABLE
+            ),
+        )
+
+
+@dataclass
+class _DummyPayloadSerializer:
+    data_converter: temporalio.converter.DataConverter
+    payload: temporalio.api.common.v1.Payload
+
+    async def serialize(self, value: Any) -> nexusrpc.Content:
+        raise NotImplementedError(
+            "The serialize method of the Serializer is not used by handlers"
+        )
+
+    async def deserialize(
+        self,
+        content: nexusrpc.Content,
+        as_type: Optional[Type[Any]] = None,
+    ) -> Any:
+        try:
+            [input] = await self.data_converter.decode(
+                [self.payload],
+                type_hints=[as_type] if as_type else None,
+            )
+            return input
+        except Exception as err:
+            raise nexusrpc.HandlerError(
+                "Data converter failed to decode Nexus operation input",
+                type=nexusrpc.HandlerErrorType.BAD_REQUEST,
+                retryable=False,
+            ) from err
+
+
+# TODO(nexus-prerelease): tests for this function
+def _exception_to_handler_error(err: BaseException) -> nexusrpc.HandlerError:
+    # Based on sdk-typescript's convertKnownErrors:
+    # https://github.com/temporalio/sdk-typescript/blob/nexus/packages/worker/src/nexus.ts
+    if isinstance(err, nexusrpc.HandlerError):
+        return err
+    elif isinstance(err, ApplicationError):
+        handler_err = nexusrpc.HandlerError(
+            # TODO(nexus-prerelease): what should message be?
+            err.message,
```

Comment:
Yeah, we will figure this out. I will try to find time next week to spec it out.